proxy/proxy.go

```diff
@@ -224,121 +215,60 @@ func localSocketPathFromUpstream(upstream string, database int, readonly bool, p
 }
 
 func (p *Proxy) ensureListenerForUpstream(upstream, originalCmd string) {
-	p.log.Info("ensuring we have a listener for", zap.String("upstream", upstream), zap.String("command", originalCmd))
+	log := p.log.With(zap.String("upstream", upstream), zap.String("command", originalCmd))
+	log.Info("ensuring we have a listener for")
+
 	p.listenerLock.Lock()
 	defer p.listenerLock.Unlock()
-	_, ok := p.listeners[upstream]
+
+	cfg, ok := p.lookup.ConfigByName(p.config.Target)
+
 	if !ok {
-		local := localSocketPathFromUpstream(upstream, p.database, p.readonly, p.config.LocalSocketPrefix, p.config.LocalSocketSuffix)
-		p.log.Info("did not find listener, creating new one", zap.String("upstream", upstream), zap.String("local", local), zap.String("command", originalCmd))
-		l, err := p.createListener(local, upstream)
-		if err != nil {
-			p.log.Error("unable to create listener", zap.Error(err))
-		}
-		p.listeners[upstream] = l
-		p.runListener(l)
+		log.Error("failed to find upstream", zap.String("target", p.config.Target))
```

Comment:
We need to re-think this. It's is going to break all of our connections to Redis cluster.

Reply:
You can run `docker compose up` in the root directory of this project and then `make test` to see the tests failing against a cluster.

Reply:
When users configure an upstream for a redis cluster they only give us one url/port. This is how every redis client works, clients are expected to issue `CLUSTER NODES` to find all of the shards in the cluster and `CLUSTER SLOTS` to map keys to which shard they should talk to.

In redisbetween we catch the `CLUSTER NODES` and `CLUSTER SLOTS` command responses at runtime and ensure that we have an upstream for every shard that they return.

Beyond that, sometimes Redis clusters get resized while an application is running. When a node gets added  redis will return a `MOVED` response with the new shard information when we try to send a command for a key that the new shard owns. We need to be able to dynamically add a listener for that case too.

Reply:
I think what you are referring to is handled at https://github.com/anujjamwal/redisbetween/blob/831dfe43e5fb5eec0acb2e32b8ff5e1ba5b07f77/proxy/proxy.go#L192
The existing behaviour is maintained where a new listener and upstream get added for every new discovered upstream from clusterNode and Moved responses.

Reply:
> You can run `docker compose up` in the root directory of this project and then `make test` to see the tests failing against a cluster.

All the tests are passing. There is one flaky test "timesout after the first command with nonzero timeout" in blocker_test.go. To be honest, I do not fully understand the intent of this test, it seems to be testing timeout behaviour of "github.com/go-redis/redis/v8" library.

Reply:
To close the loop here: Anuj explained this to me offline, I missed that `p.config.Target` is just a label referring to the cluster and not an upstream address. So we are dynamically adding upstreams here and this should work 👍